apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,553 @@
+                    .or_insert_with(|| DurationHistogram::new(None));
+                dh.increment_duration(span.end_time.duration_since(span.start_time).unwrap(), 1);
+            }
+        }
+
+        // Guarantee that the reporter is initialised
+        self.reporter
+            .get_or_try_init(|| async {
+                Reporter::try_new(self.collector.clone())
+                    .await
+                    .map_err::<ApolloError, _>(Into::into)
+            })
+            .await?;
+        let reporter = self.reporter.get_mut().unwrap();
+        // Convert the configuration data into a reportable form
+        let graph: ReporterGraph = self.graph.as_ref().unwrap().into();
+
+        // Report our consolidated statistics
+        for ((client_name, client_version, key), dh) in dh_map.drain() {
+            tracing::debug!("reporting entries: {}", dh.entries);
+            let stats = ContextualizedStats {
+                context: Some(StatsContext {
+                    client_name,
+                    client_version,
+                }),
+                query_latency_stats: Some(QueryLatencyStats {
+                    latency_count: dh.buckets,
+                    request_count: dh.entries,
+                    ..Default::default()
+                }),
+                ..Default::default()
+            };
+
+            let msg = reporter
+                .submit_stats(graph.clone(), key, stats)
+                .await
+                .map_err::<TraceError, _>(|e| e.to_string().into())?
+                .into_inner()
+                .message;
+            tracing::trace!("server response: {}", msg);
+        }
+
+        Ok(())
+    }
+}
+
+// Taken from TS implementation
+static GRAPHQL_PARSE_FAILURE: &str = "## GraphQLParseFailure\n";
+static GRAPHQL_VALIDATION_FAILURE: &str = "## GraphQLValidationFailure\n";
+static GRAPHQL_UNKNOWN_OPERATION_NAME: &str = "## GraphQLUnknownOperationName\n";
+
+fn normalize(op: Option<&opentelemetry::Value>, query: &str) -> String {
+    // If we don't have an operation name, we can't do anything useful
+    // with this query. Just return the appropriate error.
+    let op_name: String = match op {
+        Some(v) => v.as_str().into_owned(),
+        None => {
+            tracing::warn!("Could not identify operation name: {}", query);
+            return GRAPHQL_UNKNOWN_OPERATION_NAME.to_string();
+        }
+    };
+
+    let parser = Parser::new(query);
+    // compress *before* parsing to modify whitespaces/comments
+    let ast = parser.compress().parse();
+    tracing::debug!("ast:\n {:?}", ast);
+    // If we can't parse the query, we definitely can't normalize it, so
+    // just return the appropriate error.
+    if ast.errors().len() > 0 {
+        tracing::warn!("Could not parse query: {}", query);
+        return GRAPHQL_PARSE_FAILURE.to_string();
+    }
+    let doc = ast.document();
+    tracing::debug!("{}", doc.format());
```

Comment:
the call to `format()` will be done even if the level is not debug, is that call costly?

Reply:
Probably... Should I remove the debug line?

Reply:
yup, that would be a good idea

Reply:
done: b9ed1b8710c1638b4107db00ab3a53cf9bb55778